File: client/lib/services/default_data_service.dart
import 'package:flutter/foundation.dart';
import '../models/flashcard_set.dart';
import '../models/flashcard.dart';
import '../models/interview_question.dart';
import 'enhanced_http_client_service.dart';
import 'enhanced_cache_manager.dart';
import 'simple_error_handler.dart';

class DefaultDataService extends ChangeNotifier {
  static final DefaultDataService _instance = DefaultDataService._internal();
  factory DefaultDataService() => _instance;
  DefaultDataService._internal();

  final EnhancedHttpClientService _httpClient = EnhancedHttpClientService();
  final EnhancedCacheManager _cache = EnhancedCacheManager();

  Future<List<FlashcardSet>> loadDefaultFlashcardSets({String? userId}) async {
    return await SimpleErrorHandler.safe<List<FlashcardSet>>(
      () async {
        final cachedData = await _cache.getCachedData('default_flashcard_sets');
        if (cachedData != null && cachedData['flashcard_sets'] != null) {
          return _parseFlashcardSets(cachedData['flashcard_sets']);
        }

        final queryParameters = userId != null ? {'user_id': userId} : null;
        final response = await _httpClient.get('/api/default-data/flashcard-sets', queryParameters: queryParameters);

        if (response.statusCode == 200) {
          final data = response.data;
          await _cache.cacheData('default_flashcard_sets', {'flashcard_sets': data});
          return _parseFlashcardSets(data);
        }
        return <FlashcardSet>[];
      },
      fallback: <FlashcardSet>[],
      operationName: 'load_default_flashcard_sets',
    );
  }

  Future<List<InterviewQuestion>> loadDefaultInterviewQuestions({
    String? userId, String? category, String? difficulty}) async {
    return await SimpleErrorHandler.safe<List<InterviewQuestion>>(
      () async {
        final cacheKey = 'interview_${category ?? 'all'}_${difficulty ?? 'all'}';
        final cachedData = await _cache.getCachedData(cacheKey);
        if (cachedData != null && cachedData['data'] != null) {
          return _parseInterviewQuestions(cachedData['data']);
        }

        final queryParams = <String, String>{};
        if (userId != null) queryParams['user_id'] = userId;
        if (category != null) queryParams['category'] = category;
        if (difficulty != null) queryParams['difficulty'] = difficulty;

        final response = await _httpClient.get('/api/default-data/interview-questions',
            queryParameters: queryParams.isNotEmpty ? queryParams : null);

        if (response.statusCode == 200) {
          final data = response.data;
          await _cache.cacheData(cacheKey, {'data': data});
          return _parseInterviewQuestions(data);
        }
        return <InterviewQuestion>[];
      },
      fallback: <InterviewQuestion>[],
      operationName: 'load_default_interview_questions',
    );
  }

  Future<List<Map<String, dynamic>>> loadDefaultCategories() async {
    return await SimpleErrorHandler.safe<List<Map<String, dynamic>>>(
      () async {
        final cachedData = await _cache.getCachedData('categories');
        if (cachedData != null && cachedData['data'] != null) {
          return List<Map<String, dynamic>>.from(cachedData['data']);
        }

        final response = await _httpClient.get('/api/default-data/categories');
        if (response.statusCode == 200) {
          final data = response.data;
          await _cache.cacheData('categories', {'data': data});
          return List<Map<String, dynamic>>.from(data);
        }
        return <Map<String, dynamic>>[];
      },
      fallback: <Map<String, dynamic>>[],
      operationName: 'load_default_categories',
    );
  }

  Future<Map<String, int>> loadCategoryCounts({String? userId}) async {
    return await SimpleErrorHandler.safe<Map<String, int>>(
      () async {
        final cachedData = await _cache.getCachedData('category_counts');
        if (cachedData != null && cachedData['counts'] != null) {
          return Map<String, int>.from(cachedData['counts']);
        }

        final queryParameters = userId != null ? {'user_id': userId} : null;
        final response = await _httpClient.get('/api/default-data/category-counts', queryParameters: queryParameters);

        if (response.statusCode == 200) {
          final data = response.data;
          final counts = Map<String, int>.from(data['counts']);
          await _cache.cacheData('category_counts', {'counts': counts});
          return counts;
        }
        return _getDefaultCounts();
      },
      fallback: _getDefaultCounts(),
      operationName: 'load_category_counts',
    );
  }

  Map<String, int> _getDefaultCounts() => {
    'Data Analysis': 18, 'Web Development': 15, 'Machine Learning': 22, 
    'SQL': 10, 'Python': 14, 'Statistics': 8
  };

  List<FlashcardSet> _parseFlashcardSets(List<dynamic> data) {
    return data.map((setData) {
      final flashcards = (setData['flashcards'] as List<dynamic>)
          .map((cardData) => Flashcard(
                id: cardData['id'], question: cardData['question'],
                answer: cardData['answer'], isCompleted: cardData['is_completed'] ?? false,
              )).toList();
      return FlashcardSet(
        id: setData['id'], title: setData['title'], description: setData['description'],
        isDraft: setData['is_draft'] ?? false, rating: (setData['rating'] ?? 0.0).toDouble(),
        ratingCount: setData['rating_count'] ?? 0, flashcards: flashcards,
      );
    }).toList();
  }

  List<InterviewQuestion> _parseInterviewQuestions(List<dynamic> data) {
    return data.map((q) => InterviewQuestion(
          id: q['id'], 
          text: q['text'], 
          category: q['category'], 
          subtopic: q['subtopic'],
          difficulty: q['difficulty'], 
          answer: q['answer'], 
          categoryId: q['category_id'], // ✅ ADDED: Parse category_id from server
          isStarred: q['is_starred'] ?? false,
          isCompleted: q['is_completed'] ?? false, 
          isDraft: q['is_draft'] ?? false,
        )).toList();
  }

  Future<void> clearCache() async {
    await _cache.clearCache(key: 'default_flashcard_sets');
    await _cache.clearCache(key: 'category_counts');
    await _cache.clearCache(key: 'categories');
  }
}

File: server/src/services/default_data_service.py
"""
Service for providing default application data.
This service converts hardcoded client data into server-provided responses
while maintaining compatibility with future Supabase integration.
"""
import logging
from typing import List, Dict, Any, Optional
from datetime import datetime

from ..models.default_data import (
    DefaultFlashcardSetResponse,
    DefaultFlashcardResponse,
    DefaultInterviewQuestionResponse,
    DefaultCategoryResponse,
    CategoryCountsResponse
)

logger = logging.getLogger(__name__)

class DefaultDataService:
    """Service for managing default application data."""

    def __init__(self):
        """Initialize the default data service."""
        logger.info("DefaultDataService initialized")




    async def get_default_flashcard_sets(self, user_id: Optional[str] = None) -> List[DefaultFlashcardSetResponse]:
        """Get default flashcard sets - simplified version."""
        logger.info(f"Generating default flashcard sets for user_id: {user_id}")
        
        # Create a simple Python Basics set
        flashcards = []
        questions = [
            ("1", "What is Python?", "Python is a high-level programming language."),
            ("2", "How do you print in Python?", 'print("Hello World")'),
            ("3", "How do you comment in Python?", "Use # for single line comments"),
        ]
        
        for qid, q, a in questions:
            flashcards.append(DefaultFlashcardResponse(
                id=qid, question=q, answer=a, is_completed=False,
                user_id=user_id, category_id="python"
            ))
        
        sets = [DefaultFlashcardSetResponse(
            id="python-basics-001",
            title="Python Basics",
            description="",  # Empty description
            is_draft=False,
            rating=4.5,
            rating_count=12,
            flashcards=flashcards,
            user_id=user_id,
            category_id="python"
        )]
        
        return sets

    async def get_default_interview_questions(
        self, user_id: Optional[str] = None, category: Optional[str] = None, 
        difficulty: Optional[str] = None
    ) -> List[DefaultInterviewQuestionResponse]:
        """Get exactly 6 subtopics with 3 questions each (18 total questions)."""
        logger.info(f"Generating interview questions for user: {user_id}")
        
        # UPDATED: Exactly 6 subtopics with 3 questions each = 18 total questions
        questions = [
            # Data Analysis Subtopic (3 questions)
            DefaultInterviewQuestionResponse(
                id="data-analysis-1", text="How would you handle missing data in a dataset?",
                category="applied", subtopic="Data Analysis", difficulty="entry",
                answer="Identify patterns, evaluate extent, choose imputation strategy, validate approach.",
                user_id=user_id, category_id="data_analysis"
            ),
            DefaultInterviewQuestionResponse(
                id="data-analysis-2", text="Explain the difference between correlation and causation",
                category="applied", subtopic="Data Analysis", difficulty="entry",
                answer="Correlation indicates relationship strength, causation implies direct cause-effect relationship.",
                user_id=user_id, category_id="data_analysis"
            ),
            DefaultInterviewQuestionResponse(
                id="data-analysis-3", text="How do you detect outliers in data?",
                category="applied", subtopic="Data Analysis", difficulty="mid",
                answer="Use statistical methods like IQR, Z-score, or visualization techniques like box plots.",
                user_id=user_id, category_id="data_analysis"
            ),

            # Machine Learning Subtopic (3 questions)
            DefaultInterviewQuestionResponse(
                id="ml-1", text="Explain machine learning bias and variance",
                category="technical", subtopic="Machine Learning", difficulty="mid",
                answer="Bias is error from oversimplification, variance is error from sensitivity to data.",
                user_id=user_id, category_id="machine_learning"
            ),
            DefaultInterviewQuestionResponse(
                id="ml-2", text="What is overfitting and how do you prevent it?",
                category="technical", subtopic="Machine Learning", difficulty="mid",
                answer="Overfitting occurs when model memorizes training data. Prevent with regularization, cross-validation.",
                user_id=user_id, category_id="machine_learning"
            ),
            DefaultInterviewQuestionResponse(
                id="ml-3", text="Explain the difference between supervised and unsupervised learning",
                category="technical", subtopic="Machine Learning", difficulty="entry",
                answer="Supervised uses labeled data for prediction, unsupervised finds patterns in unlabeled data.",
                user_id=user_id, category_id="machine_learning"
            ),

            # SQL Database Subtopic (3 questions)
            DefaultInterviewQuestionResponse(
                id="sql-1", text="Explain the difference between INNER JOIN and LEFT JOIN",
                category="technical", subtopic="SQL Database", difficulty="entry",
                answer="INNER JOIN returns only matching records, LEFT JOIN returns all left table records.",
                user_id=user_id, category_id="sql"
            ),
            DefaultInterviewQuestionResponse(
                id="sql-2", text="What is a subquery and when would you use it?",
                category="technical", subtopic="SQL Database", difficulty="mid",
                answer="Query nested inside another query, useful for complex filtering and calculations.",
                user_id=user_id, category_id="sql"
            ),
            DefaultInterviewQuestionResponse(
                id="sql-3", text="How do you optimize a slow SQL query?",
                category="technical", subtopic="SQL Database", difficulty="senior",
                answer="Add indexes, analyze execution plan, rewrite joins, limit result sets, update statistics.",
                user_id=user_id, category_id="sql"
            ),

            # Python Programming Subtopic (3 questions)
            DefaultInterviewQuestionResponse(
                id="python-1", text="What are Python decorators?",
                category="technical", subtopic="Python Programming", difficulty="mid",
                answer="Functions that modify or extend behavior of other functions without changing their code.",
                user_id=user_id, category_id="python"
            ),
            DefaultInterviewQuestionResponse(
                id="python-2", text="Explain list comprehensions in Python",
                category="technical", subtopic="Python Programming", difficulty="entry",
                answer="Concise way to create lists using syntax: [expression for item in iterable if condition]",
                user_id=user_id, category_id="python"
            ),
            DefaultInterviewQuestionResponse(
                id="python-3", text="What is the difference between == and is in Python?",
                category="technical", subtopic="Python Programming", difficulty="entry",
                answer="== compares values, 'is' compares object identity (memory location).",
                user_id=user_id, category_id="python"
            ),

            # API Development Subtopic (3 questions)
            DefaultInterviewQuestionResponse(
                id="web-1", text="What is RESTful API design?",
                category="technical", subtopic="API Development", difficulty="mid",
                answer="Architectural style using HTTP methods for stateless, scalable web services.",
                user_id=user_id, category_id="web_development"
            ),
            DefaultInterviewQuestionResponse(
                id="web-2", text="Explain the difference between GET and POST requests",
                category="technical", subtopic="API Development", difficulty="entry",
                answer="GET retrieves data (idempotent), POST sends data to server (may cause changes).",
                user_id=user_id, category_id="web_development"
            ),
            DefaultInterviewQuestionResponse(
                id="web-3", text="What is CORS and why is it important?",
                category="technical", subtopic="API Development", difficulty="mid",
                answer="Cross-Origin Resource Sharing - security feature that controls cross-domain requests.",
                user_id=user_id, category_id="web_development"
            ),

            # Statistics Subtopic (3 questions)
            DefaultInterviewQuestionResponse(
                id="stats-1", text="What is the central limit theorem?",
                category="technical", subtopic="Statistics", difficulty="mid",
                answer="Sample means approach normal distribution as sample size increases, regardless of population distribution.",
                user_id=user_id, category_id="statistics"
            ),
            DefaultInterviewQuestionResponse(
                id="stats-2", text="Explain Type I and Type II errors",
                category="technical", subtopic="Statistics", difficulty="mid",
                answer="Type I: false positive (reject true null), Type II: false negative (accept false null).",
                user_id=user_id, category_id="statistics"
            ),
            DefaultInterviewQuestionResponse(
                id="stats-3", text="What is p-value and how do you interpret it?",
                category="technical", subtopic="Statistics", difficulty="entry",
                answer="Probability of observing results at least as extreme as observed, assuming null hypothesis is true.",
                user_id=user_id, category_id="statistics"
            ),
        ]
        
        # Apply filters
        if category:
            questions = [q for q in questions if q.category == category]
        if difficulty:
            questions = [q for q in questions if q.difficulty == difficulty]
            
        return questions
    async def get_default_categories(self) -> List[DefaultCategoryResponse]:
        """Generate categories dynamically from available questions - TRULY DYNAMIC."""
        logger.info("Generating categories dynamically from question data")
        
        # Get all questions to derive categories from actual data
        questions = await self.get_default_interview_questions()
        
        # Extract unique categories from questions
        category_data = {}
        for question in questions:
            # Use category_id as the primary category identifier for UI
            ui_category = getattr(question, 'category_id', question.category)
            if ui_category not in category_data:
                category_data[ui_category] = {
                    'subtopics': set(),
                    'question_count': 0,
                    'difficulties': set()
                }
            category_data[ui_category]['subtopics'].add(question.subtopic)
            category_data[ui_category]['question_count'] += 1
            category_data[ui_category]['difficulties'].add(question.difficulty)
        
        # Category display name mapping
        display_names = {
            'data_analysis': 'Data Analysis',
            'machine_learning': 'Machine Learning', 
            'sql': 'SQL',
            'python': 'Python',
            'web_development': 'Web Development',
            'statistics': 'Statistics',
            'technical': 'Technical Knowledge',
            'applied': 'Applied Skills',
            'behavioral': 'Behavioral Questions',
            'case': 'Case Studies',
            'job': 'Job-Specific'
        }
        
        # Generate category responses with proper metadata
        categories = []
        for category_id, data in category_data.items():
            display_name = display_names.get(category_id, category_id.replace('_', ' ').title())
            categories.append(DefaultCategoryResponse(
                id=category_id,
                name=display_name,
                subtopics=list(data['subtopics'])
            ))
        
        logger.info(f"Generated {len(categories)} categories from {len(questions)} questions")
        return categories

    async def get_category_counts(self, user_id: Optional[str] = None) -> CategoryCountsResponse:
        """Calculate category counts from actual question data - TRULY DYNAMIC."""
        logger.info(f"Calculating dynamic category counts for user: {user_id}")
        
        # Get all available questions
        questions = await self.get_default_interview_questions(user_id=user_id)
        
        # Count questions by UI category (using category_id field)
        counts = {}
        ui_category_mapping = {
            'data_analysis': 'Data Analysis',
            'machine_learning': 'Machine Learning',
            'sql': 'SQL', 
            'python': 'Python',
            'web_development': 'Web Development',
            'statistics': 'Statistics',
            # Legacy support for existing categories
            'technical': 'Technical Knowledge',
            'applied': 'Applied Skills',
            'behavioral': 'Behavioral Questions',
            'case': 'Case Studies',
            'job': 'Job-Specific'
        }
        
        for question in questions:
            # Use category_id as primary, fallback to category
            category_key = getattr(question, 'category_id', question.category)
            ui_category = ui_category_mapping.get(category_key, category_key.replace('_', ' ').title())
            counts[ui_category] = counts.get(ui_category, 0) + 1
        
        # Ensure all expected UI categories are present (even with 0 count)
        expected_categories = [
            'Data Analysis', 'Web Development', 'Machine Learning', 
            'SQL', 'Python', 'Statistics'
        ]
        for ui_cat in expected_categories:
            if ui_cat not in counts:
                counts[ui_cat] = 0
        
        total_questions = sum(counts.values())
        logger.info(f"Dynamic count calculation: {counts}, total: {total_questions}")
        
        return CategoryCountsResponse(
            counts=counts,
            total_questions=total_questions,
            user_id=user_id,
            last_updated=datetime.now()
        )
